pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -465,18 +476,41 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1beta1.PipelineRun) err
 
 // runNextSchedulableTask gets the next schedulable Tasks from the dag based on the current
 // pipeline run state, and starts them
-func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
+// after all DAG tasks are done, its responsible for scheduling final tasks and start executing them
+func (c *Reconciler) runNextSchedulableTask(ctx context.Context, pr *v1beta1.PipelineRun, d *dag.Graph, dfinally *dag.Graph, pipelineState resources.PipelineRunState, as artifacts.ArtifactStorageInterface) error {
```

Comment:
this function just got a lot larger

if you are open to it, id love to see another PR that moves this function into a separate package and takes away its dependency on the reconciler so we can unit test it

Reply:
yup indeed, would love to refactor it, will work with @afrittoli on his PR #2821 which is relevant here.

Reply:
That's what I was starting to do in https://github.com/tektoncd/pipeline/pull/2821